### ISSUE TITLE
isset -f without argument returns success

### ISSUE THREAD

**Issue by tiawl** [modernish/modernish#17]
Hi, it is me again.

My goal is not to spam you. I am currently working hard on scripts to change `/bin/sh` with `modernish` so it is possible that during coming days, I will open other issues.

The following code ...
```
#! /usr/bin/env modernish
#! use safe

if isset -f
then
  putln "I found a function with an empty name"
fi
```
... prints this in my console:
```
$ ./test_isset_f.sh
I found a function with an empty name
```

I do not think if it is an expected behavior so I investigate a little bit.
I think it is coming from those lines:
```
bin/modernish
-------------------------------------

if thisshellhas --bi=typeset &&
	command typeset -f isset >/dev/null 2>&1 &&
	! command typeset -f _Msh_nonExistentFunction >/dev/null 2>&1
then
	# bash, zsh, ksh93, mksh/pdksh, yash all have a 'typeset -f' that
	# returns with exit status 0 if the function is set, 1 if not.
	_Msh_isset_f='command typeset -f "$2" >/dev/null 2>&1'
else
	# Default (POSIX): compare output of 'command -V' against that of a known function.
	_Msh_fnOutput=$(command -V 'isset') || _Msh_initExit 'internal error (_Msh_fnOutput)'
	readonly "_Msh_fnOutput=${_Msh_fnOutput#*isset}"	# remove the function itself
	_Msh_isset_f='PATH=/dev/null command -v "$2" >/dev/null || return 1
		case $(command unalias "$2" 2>/dev/null; PATH=/dev/null command -V "$2") in
		( *"${_Msh_fnOutput}" )
			;;
		( * )	return 1 ;;
		esac'
fi
```

I installed `modernish` using `/bin/sh`. I am currently working on Ubuntu 20.04, so my `/bin/sh` is pointing on `dash`.
The `typeset` builtin is not available with `dash` so the interesting part is in the `else` section.

Surprisingly:
```
$ dash -c '{ command -v; } && echo TRUE'
TRUE
```

So the `|| return 1` part of the `_Msh_isset_f='PATH=/dev/null command -v "$2" >/dev/null || return 1` line is not triggered with `dash`.

The https://pubs.opengroup.org/onlinepubs/9699919799/utilities/command.html mentioned this about the `-v` section:
```
... Otherwise, no output shall be written and the exit status shall reflect that the name was not found.
```

The `no output shall be written` part is working fine but the `exit status shell reflect that the name was not found` part is not.
So I think it is maybe a bug of my `dash` version (0.5.10.2-6).
I did not find something talking about this in the `Bugs` section of your `README.md`.
A quick fix could be to check if the parameter used with `command -v` is not empty before running it.

Could you consider to fix it ?

Thank you.

**NB**: Surprisingly `isset -f ''` does not return 0

**Comment by tiawl** [comment]
Yes, you are right, I misread the documentation, thank you for pointing this and thank you for explaining how you made choices to build `modernish`. Moral of the story: I should consider to re-read documentation before posting an issue.

Yes it is a little bit confusing and again you are right: as a new user I feel differently.

I am going to show you my point of view and I will try to explain a possible improvement.

I really think that `isset -f` without any argument, should return an "incomplete option error" because for me it is an explicit call to the `-f`unction option. In my mind a tool's behavior should not be surprising. During my professional life, I remember an old colleague of mine coding everything he could with Ruby. He never used an other programming language. Why ? He said (and he is certainly always saying) that Ruby is the only programming language whatever you are coding with, the program will always return a predictible result and prints an expecting output. When Ruby was released, The slogan used to stand out from other programming languages was: "You get what you expect" or something like that (this is a really old story, you can not expect me to be so detailed about a slogan). 

Here, I am using `isset -f` so whatever I am giving (or not) after the `-f` I am expecting that `isset` parse it as a 'function name' and consequently, returns a matching exit status.

So now how to parse short-form shell options ? Currently `isset -f` is parsed as an option but if `isset` do not find the mandatory argument for this option finally `-f` is parsed as an argument. I think there are too much intelligence in this. I do not think that it is `isset` responsability trying to guess if `-f` is an option or an argument. I think that it is user responsability to specify if `-f` is an option or an argument: the user has to be explicit to get what he expects to get.

So my suggestion is: for any short-form shell option letter _S_, user should use `isset -- -S` to specify that `-S` is an argument. If the user type `isset -S` instead, `isset` should read `-S` as an option.

**Comment by McDutchie** [comment]
> If the first parameter value of this script is `join` this program checks if `join` is a defined function. If the first parameter value of this script is empty or equal to `-u`, this program checks for short-form shell options.

`isset` currently goes by the number of arguments to decide what to do, so the above is not true if `$1` is equal to `-u`: in that case, `isset` will currently return 2 as `-u` is an invalid function name.

However, the above is true if `$1` is empty. Hence this bit in the documentation: *If you use unquoted variable expansions here, make sure they're not empty, or the shell's empty removal mechanism will cause the wrong thing to be checked (even in the [safe mode](https://github.com/modernish/modernish#user-content-use-safe)).* So, in your example script above, using `if isset -f "${1}"` will fix the ambiguity.

But that does undermine one of the purposes of the safe mode, which is to eliminate quoting headaches. Modernish functions should be compatible with empty removal as much as possible. And, currently, `isset` just isn't.

Overall I agree with your arguments. The syntax should be changed, perhaps rethought entirely. It will be an incompatible change, so best to do it ASAP as not too many people are using modernish yet.

**Comment by McDutchie** [comment]
The [`is` function](#user-content-file-type-tests) (which I designed at a later stage than `isset`) is designed in a way that is compatible with empty removal, as the basic structure is: `is` [`-option`] `subcommand` `argument`.

Since arguments always follow a subcommand and subcommands must be present, an unquoted expansion that is removed can be reliably treated as equivalent to an empty argument, eliminating the quoting headache that results from empty removal.

So perhaps the best way to fix this bug is to get rid of `isset` altogether and fold its functionality into `is` as a new `set` subcommand. The new syntax would be like this:

* `is set` *varname*: Check if a variable is set.
* `is -v set` *varname*: Id.
* `is -x set` *varname*: Check if variable is exported.
* `is -r set` *varname*: Check if variable is read-only.
* `is -f set` *funcname*: Check if a shell function is set.
* `is set -`*optionletter* (e.g. `is set -f`): Check if shell option is set.
* `is set -o` *optionname*: Check if shell option is set by long name.

**Comment by tiawl** [comment]
Yes I think your `is` function suggestion could be a better alternative than `isset`.

**Comment by McDutchie** [comment]
I will get to this eventually, I've just been too busy with other stuff.
